### PR TITLE
added new command line option heap_size_hint for greedy GC

### PR DESCRIPTION
**Pull request by rssdev10**
The changes add a new command-line option `--heap_size_hint` for triggering aggressive garbage collection.

The following code might be used for checking that option.
```julia
using Dates

const mem_size = 0x1_000_000

counter = let
    counter = 0
    start_t = now()

    while (now() - start_t) / Millisecond(1000) < 15
        counter += 1
        arr = Vector{Int32}(undef, mem_size)
        arr[rand(1:mem_size)] = rand(Int32, 1)[begin]
        # sleep(1)
    end
    counter
end

@show counter, mem_size
```

Memory consumption might be estimated with `% gtime --verbose usr/bin/julia memory_test.jl `:
```
(counter, mem_size) = (348403, 0x01000000)
	Command being timed: "usr/bin/julia memory_test.jl"
	User time (seconds): 20.86
...
	Maximum resident set size (kbytes): 440704
...
```

Use `time` for Linux and `gtime` for MacOS.

And, memory consumption after adding the hint
`% gtime --verbose usr/bin/julia --heap-size-hint=100m memory_test.jl`:
```
(counter, mem_size) = (436, 0x01000000)
	Command being timed: "usr/bin/julia --heap-size-hint=100m memory_test.jl"
	User time (seconds): 15.24
...
	Maximum resident set size (kbytes): 267024
...
```

### CODE REVIEW

File: src/gc.c [hunk 1]
```diff
@@ -589,17 +589,19 @@ static int64_t last_gc_total_bytes = 0;
...
 // under this limit, but we will go above it rather than halting.
 #ifdef _P64
+typedef uint64_t memsize_t;
 #define default_collect_interval (5600*1024*sizeof(void*))
 static size_t max_collect_interval = 1250000000UL;
 // Eventually we can expose this to the user/ci.
-static uint64_t max_total_memory = (uint64_t) 2 * 1024 * 1024 * 1024 * 1024 * 1024;
+memsize_t max_total_memory = (memsize_t) 2 * 1024 * 1024 * 1024 * 1024 * 1024;
 #else
+typedef memsize_t uint32_t;
```
Comment (oscardssmith) [inline src/gc.c:598]: `typedef uint32_t memsize_t;`

**Review by chflood — APPROVED** [review]
Looks good

File: doc/man/julia.1 [hunk 2]
```diff
@@ -216,6 +216,11 @@ expressions. It first tries to use BugReporting.jl installed in current environm
 fallbacks to the latest compatible BugReporting.jl if not. For more information, see
 --bug-report=help.
 
+.TP
+--heap-size-hint=<size>
+Forces garbage collection if memory usage is higher that value. The memory hint might be
+specified in megabytes (500M) or gigabytes (1.5G)
```
Comment (vtjnash) [inline doc/man/julia.1:222]: Why are these in multiples of 500x or 1.5x?

Comment (oscardssmith) [reply doc/man/julia.1:222]: those are just examples. It arguably should be written `specified in megabytes (e.g. 500M) or gigabytes (e.g. 1.5G)`

Comment (rssdev10) [reply doc/man/julia.1:222]: > Why are these in multiples of 500x or 1.5x?

`long double` value is used for the numeric part. With 1e-7 restriction as a minimal value. So, even 0.000123G is working.

**Comment by MilesCranmer** [comment]
Is there an environment variable for this? Would be useful for GitHub actions

**Comment by sefffal** [comment]
I second this: I would like to set a heap size hint via environment variable for a Julia library embedded in Python.

**Comment by MilesCranmer** [comment]
^Trying to add this here: https://github.com/JuliaLang/julia/pull/55631

**Pull request merged**